File: cpp/Source/Algorithms/1.two-sum.cpp
/*
 * @lc app=leetcode id=1 lang=cpp
 *
 * [1] Two Sum
 */

// @lc code=start
#include <vector>
#include <unordered_map>

class Solution
{
public:
    std::vector<int> twoSum(std::vector<int> &nums, int target)
    {
        std::unordered_map<int, int> numToIndex;
        numToIndex[nums[0]] = 0;
        for (int i = 1; i < nums.size(); i++)
        {
            int complement = target - nums[i];
            if (numToIndex.find(complement) != numToIndex.end())
            {
                return {numToIndex[complement], i};
            }
            numToIndex[nums[i]] = i;
        }
        return {};
    };
};

// @lc code=end

File: javascript/src/1.two-sum/1.two-sum.js
/*
 * @lc app=leetcode id=1 lang=javascript
 *
 * [1] Two Sum
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

const twoSum = function (nums, target) {
    let map = new Map()
    for (let i = 0; i < nums.length; i++) {
        let complement = target - nums[i]
        if (map.has(complement)) {
            return [map.get(complement), i]
        }
        map.set(nums[i], i) // [{ nums[i], i }, ..]
    }
    // there is no solution case => return null
    return null
};

module.exports = twoSum;

// @lc code=end

